cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java

```diff
@@ -78,24 +85,61 @@ public InMemoryDatasetFramework(DatasetDefinitionRegistryFactory registryFactory
   public InMemoryDatasetFramework(DatasetDefinitionRegistryFactory registryFactory,
                                   @Named("defaultDatasetModules") Map<String, ? extends DatasetModule> defaultModules) {
     this.registryFactory = registryFactory;
-    this.defaultModules = defaultModules;
-    resetRegistry();
+    this.namespaces = Sets.newHashSet();
+    this.nonDefaultTypes = HashMultimap.create();
+    this.instances = HashBasedTable.create();
+    this.registries = Maps.newHashMap();
+    // the order in which module classes are inserted is important,
+    // so we use a table where Map<Id.DatasetModule, String> is a LinkedHashMap
+    Map<Id.Namespace, Map<Id.DatasetModule, String>> backingMap = Maps.newHashMap();
+    this.moduleClasses = Tables.newCustomTable(backingMap, new Supplier<Map<Id.DatasetModule, String>>() {
+      @Override
+      public Map<Id.DatasetModule, String> get() {
+        return Maps.newLinkedHashMap();
+      }
+    });
+
+    // add default dataset modules to system namespace
+    namespaces.add(Constants.SYSTEM_NAMESPACE_ID);
+    DatasetDefinitionRegistry systemRegistry = registryFactory.create();
+    for (Map.Entry<String, ? extends DatasetModule> entry : defaultModules.entrySet()) {
+      LOG.info("Adding Default module {} to system namespace", entry.getKey());
+      String moduleName = entry.getKey();
+      DatasetModule module = entry.getValue();
+      entry.getValue().register(systemRegistry);
+      // keep track of default module classes. These are used when creating registries for other namespaces,
+      // which need to register system classes too.
+      String moduleClassName = DatasetModules.getDatasetModuleClass(module).getName();
+      Id.DatasetModule moduleId = Id.DatasetModule.from(Constants.SYSTEM_NAMESPACE, moduleName);
```

Comment:
nit: Perhaps use ``Constants.SYSTEM_NAMESPACE_ID``, so we don't create another ``Id.Namespace``